core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/IdempotentAsyncResponseHandler.java

```diff
@@ -0,0 +1,117 @@
+ * per state change and will cache and reserve the future that is returned by the delegate the rest of the time.
+ *
+ * One application of this wrapper is to ensure that prepare() is not called on the underlying response handler more
+ * than once per retry no matter where or how many times its invoked.
+ *
+ * This class is asserted to be thread-safe and it should be fine to have multiple threads call prepare()
+ * simultaneously.
+ *
+ * @param <T> The type of the wrapped {@link TransformingAsyncResponseHandler}
+ * @param <R> The type of the object used to determine scope.
+ */
+@SdkInternalApi
+@ThreadSafe
+public class IdempotentAsyncResponseHandler<T, R> implements TransformingAsyncResponseHandler<T> {
+    private final Supplier<R> newScopeSupplier;
+    private final BiPredicate<R, R> newScopeInRangePredicate;
+    private final TransformingAsyncResponseHandler<T> wrappedHandler;
+
+    // Taking advantage of the volatile properties of AtomicReference
+    private final AtomicReference<R> cachedScope = new AtomicReference<>();
+    private final AtomicReference<CompletableFuture<T>> cachedPreparedFuture = new AtomicReference<>();
+
+    private IdempotentAsyncResponseHandler(TransformingAsyncResponseHandler<T> wrappedHandler,
+                                           Supplier<R> newScopeSupplier,
+                                           BiPredicate<R, R> newScopeInRangePredicate) {
+        this.newScopeSupplier = newScopeSupplier;
+        this.newScopeInRangePredicate = newScopeInRangePredicate;
+        this.wrappedHandler = wrappedHandler;
+    }
+
+    /**
+     * Creates a new wrapped {@link TransformingAsyncResponseHandler}
+     * @param wrappedHandler A different handler to wrap.
+     * @param preparedScopeSupplier A function that retrieves the current scope to determine if a new future can be
+     *                              prepared.
+     * @param scopeInRangePredicate A predicate that can be used to test the cached scope against the newly retrieved
+     *                             scope to determine if a new future can be prepared.
+     * @param <T> The type of the wrapped {@link TransformingAsyncResponseHandler}
+     * @param <R> The type of the object used to determine scope.
+     */
+    public static <T, R> IdempotentAsyncResponseHandler<T, R> create(
+            TransformingAsyncResponseHandler<T> wrappedHandler,
+            Supplier<R> preparedScopeSupplier,
+            BiPredicate<R, R> scopeInRangePredicate) {
+
+        return new IdempotentAsyncResponseHandler<>(wrappedHandler,
+                                                    preparedScopeSupplier,
+                                                    scopeInRangePredicate);
+    }
+
+    @Override
+    public CompletableFuture<T> prepare() {
+        // Test if the new scope is in range of the cached scope; if not delegate and cache result and new scope
+        if (cachedPreparedFuture.get() == null ||
```

Comment:
This is some heavy new code added along a common path...

Reply:
It's replacing some code that instead persists the CompletableFuture object created by prepare() in ExecutionAttributes and then reads it out later when deciding whether to call prepare a second time. Every retry also writes explicitly writes null into ExecutionAttributes to effectively 'reset' the state. I think this replacement wrapper is not heavier than the code it replaces.

Reply:
For reference the implementation of .get() on an AtomicReference just reads and returns the value of a volatile member variable.

Reply:
As discussed leaving as-is for now, but considering as a candidate for attention when tuning for performance. An option we have is to remove the thread-safety, which will give us a small gain on prepare()